### PR TITLE
fix getPackageInfo deprecation

### PR DESCRIPTION
**Pull request by david-allison**
## Purpose / Description

Deprecated in API 33: https://developer.android.com/reference/android/content/pm/PackageManager#getPackageInfo(java.lang.String,%20int) 

I'm going to use this functionality in Scoped Storage, so I've prioritised fixing this deprecation. 

## Fixes
Fixes https://github.com/ankidroid/Anki-Android/pull/12548 (superseded)
  * Sincere & profuse apologies to @n1snt - I'm marking a concerted effort to get back into both maintenance and development to ensure this doesn't occur again.
  
Related https://github.com/ankidroid/Anki-Android/issues/12364

## Approach
⚠️ I'm unsure on this approach, and welcome alternate suggestions

`flags: Integer` was replaced with `flags: PackageInfoFlags` in API 33. This is likely because the `int` flags were expanded to be `long` flags.

In addition to this, an interface with a `@LongDef` was added to the constructor of `PackageInfoFlags` to allow for safer construction.

We define our own `PackageInfoFlags`, then `PackageInfoFlagsBits`, then reproduce the constants which are unavailable in lesser versions of the API. This **should** mean that `PackageInfoFlags` can be used by API 21 without issue and matches the API to the closest of our ability

I say **should** as for all the implementations I've seen in the Android codebase. 

Flags are tested using the standard `bit & flag == 0`, which is forwards-compatible with the addition of unknown flags.

## How Has This Been Tested?

Unit tested

## Learning (optional, can help others)
https://developer.android.com/reference/android/content/pm/PackageManager#getPackageInfo(java.lang.String,%20int)

https://cs.android.com/android/platform/superproject/+/master:frameworks/base/services/core/java/com/android/server/pm/ComputerEngine.java;drc=c4ad8bc669e66262a00798b57132347a0d0aa2ac;bpv=1;bpt=1;l=1705?q=getPackageInfoInternal&ss=android&gsn=getPackageInfoInternalBody&gs=kythe%3A%2F%2Fandroid.googlesource.com%2Fplatform%2Fsuperproject%3Flang%3Djava%3Fpath%3Dcom.android.server.pm.ComputerEngine%23977e4a94695fef516f4b2d9fa73dea77cfaf06eff40c6fb3ec9bd80c6e18a08f

## Checklist
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/compat/PackageManagerCompat.kt [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+ * Provides [PackageInfoFlags] to Android versions before SDK 33,
+ * keeping a near consistent API for [Compat.getPackageInfo],
+ * constraining the flags to the correct values (or types when available)
+ *
+ * Allows for either long flags to be provided (new API), or int flags (old API).
+ * The old API is currently better as it means .toLong() isn't needed on constants provided
+ * to the API.
+ * For future: Can Kotlin can accept either the int or long in a `@LongDef`
+ */
+class PackageManagerCompat {
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/compat/PackageManagerCompat.kt:52]: Do we need this to be a class? I'd get confused seeing both this and the platform `PackageManagerCompat` on the IDE auto-suggestion

File: AnkiDroid/src/main/java/com/ichi2/compat/PackageManagerCompat.kt [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+    /**
+     * Specific flags used for retrieving package info. Example:
+     * `PackageManager.getPackageInfo(packageName, PackageInfoFlags.of(0)`
+     */
+    class PackageInfoFlags private constructor(@PackageInfoFlagsBits value: Long) : Flags(value) {
+        companion object {
+            fun of(@PackageInfoFlagsBits value: Long): PackageInfoFlags = PackageInfoFlags(value)
+
+            val EMPTY = PackageInfoFlags(0)
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/compat/PackageManagerCompat.kt:67]: nit: add a javadoc stating that this is a helper property and doesn't exist on the platform API

File: AnkiDroid/src/main/java/com/ichi2/compat/CompatHelper.kt [hunk 3]
```diff
@@ -83,6 +83,18 @@ class CompatHelper private constructor() {
...
+         * These Kotlin-based extensions make the Compat API cleaner.
+         *
+         * Instead of:
+         * val packageInfo = compat.getPackageInfo(context.packageManager, name, PackageInfoFlags.of(0))
+         *
+         * we have the cleaner:
+         *
+         * val packageInfo = context.packageManager.getPackageInfoCompat(name, PackageInfoFlags.of(0))
+         */
+
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/compat/CompatHelper.kt:97]: I particularly don't find this useful. But I don't care enough if it stays or not

Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/compat/CompatHelper.kt:97]: Had a few questions about why we've changed compat from the Java-based `CompatHelper.compat.`... 

Reviewer's choice

Comment (BrayanDSO) [reply AnkiDroid/src/main/java/com/ichi2/compat/CompatHelper.kt:97]: I believe that the fact that methods `CompatHelper` are helpers makes this obvious enough. The comment is almost redundant to me, so I'd vote to remove it

**Review by mikehardy — APPROVED** [review]
works for me if @BrayanDSO is +1 on it after the last push

**Comment by BrayanDSO** [comment]
I need to double check some things. Will get into it when I get to my dear PC

**Review by BrayanDSO — APPROVED** [review]
LGTM

**Pull request merged**